Specify Client Versions on Engine API (#517)
ethereum/execution-apis: src/engine/identification.md

- [Structures](#structures)
  - [ClientCode](#clientcode)
  - [ClientIdentificationV1](#clientidentificationv1)
- [Methods](#methods)
  - [engine_clientIdentificationV1](#engine_clientidentificationv1)
    - [Request](#request)
    - [Response](#response)
    - [Specification](#specification)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Client Identification

To facilitate a more accurate measurement of execution layer client diversity statistics, execution clients **SHOULD** support the methods described in this document.

## Structures

### ClientCode

This enum defines a standard for specifying a client with just two letters. Clients teams which have a code reserved in this list **MUST** use this code when identifying themselves. The code is specified here only to facilitate standardization and NOT to imply that these are the only supported Ethereum clients. Any clients not listed here are free to use any two letters which don't collide with an existing client code. They are encouraged to make a PR to this repo to reserve their own code. Existing codes are as follows:

 - `BU`: besu
 - `EJ`: ethereumJS
 - `EG`: erigon

>  - `GE`: go-ethereum

Reviewer: May as well reserve `GR` here for Grandine, as it's a client that exists regardless of its open-source status

Author: done! :)

 - `LH`: lighthouse
 - `LS`: lodestar
 - `NM`: nethermind
 - `NB`: nimbus
 - `TK`: teku
 - `PM`: prysm
 - `RH`: reth
 
### ClientIdentificationV1

This structure contains information which identifies a client implementatiopn. The fields are encoded as follows:

- `code`: `ClientCode`, e.g. `NB` or `BU`
- `clientName`: `string`, Human-readable name of the client, e.g. `Lighthouse` or `go-ethereum`

> - `version`: `string`, the standard semantic version string of the current implementation e.g. `v4.6.0` or `1.0.0-alpha.1` or `1.0.0+20130313144700`

Reviewer: Being pedantic: not all clients use semantic versioning (e.g. Teku/Besu which use Ubuntu-style date versions).

```suggestion
- `version`: `string`, the version string of the current implementation e.g. `v4.6.0` or `1.0.0-alpha.1` or `1.0.0+20130313144700`
```

Author: done! :D

## Client Version Specification

To facilitate a more accurate measurement of execution layer client diversity statistics, execution clients **SHOULD** support the methods described in this document.

## Structures

### ClientCode

This enum defines a standard for specifying a client with just two letters. Clients teams which have a code reserved in this list **MUST** use this code when identifying themselves. The code is specified here only to facilitate standardization and NOT to imply that these are the only supported Ethereum clients. Any clients not listed here are free to use any two letters which don't collide with an existing client code. They are encouraged to make a PR to this repo to reserve their own code. Existing codes are as follows:

 - `BU`: besu
 - `EJ`: ethereumJS
 - `EG`: erigon
 - `GE`: go-ethereum
 - `GR`: grandine
 - `LH`: lighthouse
 - `LS`: lodestar
 - `NM`: nethermind
 - `NB`: nimbus
 - `TK`: teku
 - `PM`: prysm
 - `RH`: reth

### ClientVersionV1

This structure contains information which identifies a client implementation. The fields are encoded as follows:

- `code`: `ClientCode`, e.g. `NB` or `BU`

> - `clientName`: `string`, Human-readable name of the client, e.g. `Lighthouse` or `go-ethereum`

Reviewer 2: ```suggestion
- `name`: `string`, Human-readable name of the client, e.g. `Lighthouse` or `go-ethereum`
```
Slight pref for "name". It's implied it is the client. There isn't "clientCode", "clientVersion", etc.

- `version`: `string`, the version string of the current implementation e.g. `v4.6.0` or `1.0.0-alpha.1` or `1.0.0+20130313144700`

> - `commit`: `string`, the hex of the first 4 bytes of the latest commit hash of this build e.g. `fa4ff922`

Reviewer 2: ```suggestion
- `commit`: `DATA`, 4 bytes - first four bytes of the latest commit hash of this build e.g. `fa4ff922`
```

Reviewer 2: probably shouldn't mix `0x`-less hex with `0x`-ed hex.
